packages/functions/src/primitive-outline.ts

```diff
@@ -0,0 +1,135 @@
+			indexArr[1] = indices[i * 3 + 1];
+			indexArr[2] = indices[i * 3 + 2];
+		} else {
+			indexArr[0] = i * 3 + 0;
+			indexArr[1] = i * 3 + 1;
+			indexArr[2] = i * 3 + 2;
+		}
+		const _triangle = new Triangle(
+			[positions[indexArr[0] * 3 + 0], positions[indexArr[0] * 3 + 1], positions[indexArr[0] * 3 + 2]] as vec3,
+			[positions[indexArr[1] * 3 + 0], positions[indexArr[1] * 3 + 1], positions[indexArr[1] * 3 + 2]] as vec3,
+			[positions[indexArr[2] * 3 + 0], positions[indexArr[2] * 3 + 1], positions[indexArr[2] * 3 + 2]] as vec3,
+		);
+		const { a, b, c } = _triangle;
+		const normal = [0, 0, 0] as vec3;
+		Triangle.getNormal(a, b, c, normal);
+
+		hashes[0] = `${Math.round(a[0] * precision)},${Math.round(a[1] * precision)},${Math.round(a[2] * precision)}`;
+		hashes[1] = `${Math.round(b[0] * precision)},${Math.round(b[1] * precision)},${Math.round(b[2] * precision)}`;
+		hashes[2] = `${Math.round(c[0] * precision)},${Math.round(c[1] * precision)},${Math.round(c[2] * precision)}`;
+
+		// skip degenerate triangles
+		if (hashes[0] === hashes[1] || hashes[1] === hashes[2] || hashes[2] === hashes[0]) {
+			continue;
+		}
+
+		for (let j = 0; j < 3; j++) {
+			const jNext = (j + 1) % 3;
+			const vecHash0 = hashes[j];
+			const vecHash1 = hashes[jNext];
+			const v0 = _triangle[vertKeys[j]];
+			const v1 = _triangle[vertKeys[jNext]];
+			const hash = `${vecHash0}_${vecHash1}`;
+			const reverseHash = `${vecHash1}_${vecHash0}`;
+			if (reverseHash in edgeData && edgeData[reverseHash]) {
+				// if we found a sibling edge add it into the vertex array if
+				// it meets the angle threshold and delete the edge from the map.
+				if (glvec3.dot(normal, edgeData[reverseHash].normal) <= thresholdAngle) {
+					vertices.push(v0[0], v0[1], v0[2]);
+					vertices.push(v1[0], v1[1], v1[2]);
+				}
+
+				edgeData[reverseHash] = null;
+			} else if (!(hash in edgeData)) {
+				// if we've already got an edge here then skip adding a new one
+				edgeData[hash] = {
+					index0: indexArr[j],
+					index1: indexArr[jNext],
+					normal: glvec3.clone(normal),
+				};
+			}
+		}
+	}
+
+	// iterate over all remaining, unmatched edges and add them to the vertex array
+	for (const key in edgeData) {
+		if (edgeData[key]) {
+			const { index0, index1 } = edgeData[key];
+			const v0 = glvec3.fromValues(
+				positions[index0 * 3 + 0],
+				positions[index0 * 3 + 1],
+				positions[index0 * 3 + 2],
+			);
+			const v1 = glvec3.fromValues(
+				positions[index1 * 3 + 0],
+				positions[index1 * 3 + 1],
+				positions[index1 * 3 + 2],
+			);
+
+			vertices.push(v0[0], v0[1], v0[2]);
+			vertices.push(v1[0], v1[1], v1[2]);
+		}
+	}
+
+	const accesor = document
+		.createAccessor()
+		.setArray(new Float32Array(vertices))
+		.setType('VEC3')
+		.setBuffer(positionAccessor.getBuffer());
+
+	const edgePrim = document
+		.createPrimitive()
+		.setName(prim.getName() + '_edges')
+		.setMode(Primitive.Mode.LINES)
+		.setAttribute('POSITION', accesor);
```

Comment:
Perhaps ideally, all vertex attributes from the source primitive would be preserved in the output, but I recognize that's a fairly tricky change. For example, so that edges of a skinned mesh can be animated along with the rest of the mesh. Feel free to skip it and we can include it as a later improvement!